# DMA Assignment 1 - Data Preprocessing

Data transformations are useful for preparing a dataset for answering a particular question. Part of this process involves generating features from the dataset you find relevant to the question at hand. For this lab, we will be using a Yelp reviews dataset. Each row in the dataset depicts one review along with the features of the review (the reviewer, the review text, etc.). The goal of this lab is to convert this reviews dataset into a reviewers dataset by creating different features of each reviewer.

The submission for this assignment should be done *individually*, although you are allowed to work in groups of 2.



sharable link: https://colab.research.google.com/drive/1Q8ANFkLFNJE0xZSzkMGSK-LNX9PcBWAS

# Environment Setup
Run this cell to setup your environment.

In [3]:
# Importing libraries
import pandas as pd
import math
import numpy as np
print('Libraries Imported')

#DOWNLOADING DATASET
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
#!unzip yelp_reviews.zip
print('Dataset Downloaded: yelp_reviews.csv')
df=pd.read_csv('yelp_reviews.csv')
print(df.head())

print('Setup Complete')

Libraries Imported
--2019-02-05 20:51:14--  http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 376638166 (359M) [text/csv]
Saving to: ‘yelp_reviews.csv’

yelp_reviews.csv    100%[===================>] 359.19M  2.47MB/s    in 6m 0s   

2019-02-05 20:57:15 (1021 KB/s) - ‘yelp_reviews.csv’ saved [376638166/376638166]

Dataset Downloaded: yelp_reviews.csv
     type             business_id                 user_id  stars  \
0  review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
1  review  mxrXVZWc6PWk81gvOVNOUw  0aN5QPhs-VwK2vusKG0waQ      5   
2  review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3  review  mxrXVZWc6PWk81gvOVNOUw  1JUwyYab-uJzEx_FRd81Zg      5   
4  review  mxrXVZWc6PWk81gvOVNOUw  2Zd3Xy8h

**Q1: What was the highest number of reviews for any one business_id?**
- For this task, we will need to group the reviews dataset by business_id. This will aggregate data for each business, which is what we need. This can be done using [groupby](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html).
    - ```yelp_businesses = yelp_dataset.groupby('business_id').size() ```
    - The .size() counts the number of instances for each business_id, which gives us the number of reviews as each instance in this dataset is a review.
    - The following command will sort this list, after which you can take note of the highest value: ```sorted_yelp_businesses = yelp_businesses.sort(ascending=False,inplace=False)```
    - This approach allows you to see the data structure being used in the sort. A quicker approach to getting the max would be to use the max function: ```max(yelp_businesses)```
<br>
<br>

In [4]:
# YOUR ANSWER HERE
yelp_businesses = df.groupby('business_id').size()

#dataframe.sort() was deprecated since developers need to differentiate between
#sort values or sort index

sorted_yelp_businesses = yelp_businesses.sort_values(
  ascending=False, inplace=False)
print("The maximum number of reviews are {}.".format(
  max(sorted_yelp_businesses)))

The maximum number of reviews are 4128.


###Answer: 4128

**Q2: What was the average number of reviews for a business_id?**

In [5]:
# YOUR CODE HERE
print(yelp_businesses.describe())
print("The average number of reviews for a business_id is {:.2f} reviews.".format(yelp_businesses.mean()))

count    43317.000000
mean        12.634139
std         63.861190
min          1.000000
25%          1.000000
50%          3.000000
75%          7.000000
max       4128.000000
dtype: float64
The average number of reviews for a business_id is 12.63 reviews.


###Answer: 12.6341

**Q3: What is the average number of reviews per reviewer?**

In [6]:
# YOUR CODE HERE
## YOUR ANSWER HERE
yelp_users = df.groupby('user_id').size()

#dataframe.sort() was deprecated since developers need to differentiate between 
#sort values or sort index
sorted_yelp_users = yelp_users.sort_values(ascending=False, inplace=False)
#print(sorted_yelp_users.head())

print("The average number of reviews per user are {:.4f}.".format(
  sorted_yelp_users.mean()))


The average number of reviews per user are 3.1885.


###Answer: 3.1885

**Q4: What is the average number of cool votes per reviewer?**

In [7]:
# YOUR CODE HERE
users_votes = df.groupby(['user_id']).sum()
users_votes['cool_votes'].describe()
print('The average number of cool votes per reviewer (user id) is {:.4f} cool reviews.'.format(
  users_votes['cool_votes'].mean()))

The average number of cool votes per reviewer (user id) is 1.2417 cool reviews.


###Answer: 1.2417

**Q5: What is the average number of funny votes per reviewer?**

In [8]:
# YOUR CODE HERE
print('The average number of funny votes per reviewer (user id) is {:.4f} funny reviews.'.format(
  users_votes['funny_votes'].mean()))

The average number of funny votes per reviewer (user id) is 1.1013 funny reviews.


###Answer: 1.1013

**Q6: What is the average number of useful votes per reviewer?**

In [9]:
# YOUR CODE HERE
print('The average number of useful votes per reviewer (user id) is {:.4f} useful reviews.'.format(
  users_votes['useful_votes'].mean()))

The average number of useful votes per reviewer (user id) is 2.4845 useful reviews.


###Answer: 2.4845

**Q7: What is the average of the log of the number of reviews per reviewer?**

In [10]:
# YOUR CODE HERE
log_reviews = np.log(sorted_yelp_users)
log_reviews.describe()
print('The average of the log of the numbers of reviews per reviewer (user_id) is {:.4} reviews.'.format(log_reviews.mean()))


The average of the log of the numbers of reviews per reviewer (user_id) is 0.7682 reviews.


###Answer: 0.7682

**Q8: What is the average of the log of the number of cool votes per reviewer?**

We have an issue performing log of 0. Therefore we can employ two strategies:
    1. Eliminate rows that contain a 0 in the votes columns.
    2. Change -inf to 0 after performing the log (this can also be done by changing all 0 values to 1. This way the log of 1 will be 0.

In [11]:
# YOUR CODE HERE
# Strategy #1: eliminate rows with 0 cool votes.
only_cool_users = users_votes[users_votes.cool_votes != 0]
only_cool_users = np.log(only_cool_users['cool_votes'])
only_cool_users.describe()
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(only_cool_users.mean()))

# Strategy #2:if we apply log to cool votes and do NOT eliminate rows but instead replace -inf for 0
cool_users_zero = users_votes.replace(to_replace = 0, value = 1)
cool_users_zero = np.log(cool_users_zero['cool_votes'])
#print(cool_users_zero.head())
#cool_users_zero.describe()
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(cool_users_zero.mean()))

The mean of the log of the number of cool votes per reviewer are 0.67 votes.
The mean of the log of the number of cool votes per reviewer are 0.21 votes.


###Answer: 0.67 or 0.21

**Q9: What is the average of the log of the number of funny votes per reviewer?**

We will employ the two same strategies used in Q8.

In [12]:
# YOUR CODE HERE
# Strategy #1
only_funny_users = users_votes[users_votes.funny_votes != 0]
only_funny_users = np.log(only_funny_users['funny_votes'])
#print(only_funny_users.describe())
#print(only_funny_users.head())
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(only_funny_users.mean()))

# Strategy #2: if we apply log to funny votes and do NOT eliminate rows but instead replace -inf for 0
funny_users_zero = users_votes.replace(to_replace = 0, value = 1)
funny_users_zero = np.log(funny_users_zero['funny_votes'])
#print(funny_users_zero.head())
#funny_users_zero.describe()
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(funny_users_zero.mean()))

The mean of the log of the number of cool votes per reviewer are 0.66 votes.
The mean of the log of the number of cool votes per reviewer are 0.19 votes.


###Answer: 0.66 or 0.19

**Q10: What is the average of the log of the number of useful votes per reviewer?**

We will employ the two same strategies used in Q8.

In [13]:
# YOUR CODE HERE
# Strategy #1
only_useful_users = users_votes[users_votes.useful_votes != 0]
only_useful_users = np.log(only_useful_users['useful_votes'])
#print(only_useful_users.describe())
#print(only_useful_users.head())
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(only_useful_users.mean()))

# Strategy #2: if we apply log to useful votes and do NOT eliminate rows but instead replace -inf for 0
useful_users_zero = users_votes.replace(to_replace = 0, value = 1)
useful_users_zero = np.log(useful_users_zero['useful_votes'])
#print(useful_users_zero.head())
#useful_users_zero.describe()
print('The mean of the log of the number of cool votes per reviewer are {:.2f} votes.'.format(useful_users_zero.mean()))

The mean of the log of the number of cool votes per reviewer are 0.91 votes.
The mean of the log of the number of cool votes per reviewer are 0.48 votes.


###Answer: 0.91 or 0.48

The following three questions can be answered differently 

**Q11: Find the average of the percentage of total cool votes out of total votes for each reviewer.**

In [0]:
percentage = df.copy()
percentage = percentage.groupby('user_id')
cool_sum = percentage['cool_votes'].sum()
funny_sum = percentage['funny_votes'].sum()
useful_sum = percentage['useful_votes'].sum()
total = cool_sum + funny_sum + useful_sum

In [29]:
cool_percentage = cool_sum/total
print('The average of the percentage of total cool votes out of total votes for each reviewer is {:.2f} or {:0.2f}% without filing N/A values.'.format(
    cool_percentage.mean(), cool_percentage.mean() * 100))
cool_percentagenafilled = cool_percentage.fillna(value = 0)
print('The average of the percentage of total cool votes out of total votes for each reviewer is {:.2f} or {:0.2f}% after filling N/A values.'.format(
    cool_percentagenafilled.mean(), cool_percentagenafilled.mean() * 100))

The average of the percentage of tootal cool votes out of total votes for each reviewer is 0.19 or 19.27% without filing N/A values.
The average of the percentage of tootal cool votes out of total votes for each reviewer is 0.11 or 11.27% after filling N/A values.


###Answer: 19.27% without filling N/A values and 11.27% if we fill N/A values

**Q12: Find the average of the percentage of total funny votes out of total votes for each reviewer.**

In [30]:
funny_percentage = funny_sum/total
print('The average of the percentage of total funny votes out of total votes for each reviewer is {:.2f} or {:0.2f}% without filing N/A values.'.format(
    funny_percentage.mean(), funny_percentage.mean() * 100))
funny_percentagenafilled = funny_percentage.fillna(value = 0)
print('The average of the percentage of total funny votes out of total votes for each reviewer is {:.2f} or {:0.2f}% after filling N/A values.'.format(
    funny_percentagenafilled.mean(), funny_percentagenafilled.mean() * 100))

The average of the percentage of total funny votes out of total votes for each reviewer is 0.18 or 18.26% without filing N/A values.
The average of the percentage of total funny votes out of total votes for each reviewer is 0.11 or 10.67% after filling N/A values.


###Answer: 18.26% without filling N/A values and 10.67% if we fill N/A values

**Q13: Find the average of the percentage of total useful votes out of total votes for each reviewer.**

In [31]:
useful_percentage = useful_sum/total
print('The average of the percentage of total useful votes out of total votes for each reviewer is {:.2f} or {:0.2f}% without filing N/A values.'.format(
    useful_percentage.mean(), useful_percentage.mean() * 100))
useful_percentagenafilled = useful_percentage.fillna(value = 0)
print('The average of the percentage of total useful votes out of total votes for each reviewer is {:.2f} or {:0.2f}% after filling N/A values.'.format(
    useful_percentagenafilled.mean(), useful_percentagenafilled.mean() * 100))

The average of the percentage of total useful votes out of total votes for each reviewer is 0.62 or 62.47% without filing N/A values.
The average of the percentage of total useful votes out of total votes for each reviewer is 0.37 or 36.53% after filling N/A values.


###Answer: 62.47% without filling N/A values and 36.53% if we fill N/A values

**Q14: Average review text length (in non-space characters)**

In [32]:
# YOUR CODE HERE
text_length = df['text'].str.len() - df['text'].str.count(' ')
#print(text_length.head())
#print(text_length.describe())
print('The average review text length (characters) not counting spaces are {:.0f} characters.'.format(text_length.mean()))

The average review text length (characters) not counting spaces are 499 characters.


###Answer: 499 characters

**Q15: Year in which the reviewer wrote the most reviews. Once you have this for each reviewer, subtract the minimum possible year (2005) from each so that your final feature values are 0, 1, 2 etc.**

In [36]:
yeardf = df.copy()
yeardf['date'] = pd.to_datetime(yeardf['date'])
yeardf['year'] = yeardf['date'].dt.year
#yeardf.head()

useryearly = yeardf.groupby(['user_id', 'year']).size().to_frame('count')
maxcount = useryearly.groupby('user_id')['count'].transform(max) == useryearly['count']
maxfreq = useryearly[maxcount].reset_index()

maxfreq['finalvalue'] = maxfreq['year'] - 2005

maxfreq.head()

user_id  year  count  finalvalue
0  --1Y03CEKR3WDbBjYnsW7A  2011      1           6
1  --2QZsyXGz1OhiD4-0FQLQ  2014      3           9
2  --82_AVgRBsLw6Dhy8sEnA  2008      1           3
3  --8A9o_NeGyt_3kzlXtSdg  2015      2          10
4  --8WbseBk1NjfPiZWjQ-XQ  2016      4          11

###Answer: The finalvalue column.

**Q16: Come up with a new feature. This may be derived from existing features. Give your feature the name ** *my_new_feature.* **Dsiplay head() of this new feature.**

I will create a new column called "year" where I only extract the year after converting the date column to a dateframe. The completely new feature is the importance attribute where 0 is not important and 3 is the most important type of review. 3 means the comment is funny, useful, and cool.

In [35]:
# YOUR CODE HERE

df['date'] = pd.to_datetime(df['date'])
df['year'] = df['date'].dt.year
newdf = df.copy()
newdf['importance'] = newdf['cool_votes'] + newdf['useful_votes'] + newdf['funny_votes']
newdf['business_id'] = newdf['business_id'].astype('category')
newdf['user_id'] = newdf['user_id'].astype('category')
newdf['text_length'] = text_length

newdf[['cool_votes','useful_votes','funny_votes']] = newdf[['cool_votes','useful_votes','funny_votes']].astype(bool).astype(int)
#newdf.info()
#newdf['user_id'].value_counts()
newdf.head()


type             business_id                 user_id  stars  \
0  review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
1  review  mxrXVZWc6PWk81gvOVNOUw  0aN5QPhs-VwK2vusKG0waQ      5   
2  review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3  review  mxrXVZWc6PWk81gvOVNOUw  1JUwyYab-uJzEx_FRd81Zg      5   
4  review  mxrXVZWc6PWk81gvOVNOUw  2Zd3Xy8hUVmZkNg7RyNjhg      4   

                                                text       date  cool_votes  \
0  Definitely try the duck dish.   I rank it amon... 2011-06-13           0   
1  Big Ass Burger was awesome! Great $5 mojitos. ... 2011-06-25           1   
2             Unbelievable sandwiches! Good service. 2011-06-25           0   
3  Awesome, awesome, awesome! My mom and sister a... 2011-07-18           1   
4  I had the ribs they were great.  The beer sele... 2011-07-19           1   

   useful_votes  funny_votes  year  importance  text_length  
0             0            0  2011           0          162  
1             0            0  2011           1           84  
2             0            0  2011           0           35  
3             1            0  2011           2          257  
4             0            1  2011           2          147

###Answer: Importance Feature where 0 is not important and 3 is the most important type of review. 3 means the comment has all three attributes (funny, useful, and cool.)